accepted/2022/capability-analyzer/capability-analyzer.md

```diff
@@ -0,0 +1,205 @@
+relatively large data files to a cloud backend. To avoid disappointment down the
+road, the UI performs some upfront validation that parses the input files. To
+prevent the UI from freezing up, someone wrote some code that offloaded that
+logic to a worker thread. While that is not ideal, that worked well for the last
+decade in the .NET Framework app. She copy and pastes a good chunk of code that
+includes the following snippet:
+
+```C#
+void StartValidation()
+{
+    InputValidationConfiguration args = GetConfiguration();
+    SynchronizationContext syncContext = SynchronizationContext.Current;
+
+    Thread thread = new Thread(() =>
+    {
+        var result = Validator.Validate(args);
+        syncContext.Post(_ => ValidationComplete(result));
+    });
+    
+    thread.Start();
+}
+```
+
+She immediately gets a diagnostic:
+
+    CAXXXX: The API 'Thread.Start()' requires the capability 'ThreadCreation'. Either mark the consumer with `[RequiresThreadCreation]` or check by calling 'Thread.IsCreationSupported'.
+
+After some research into what this means she realizes that not all runtime
+targets allow creating new threads, which includes her desired target, which is
+WebAssembly. After some reading, she realizes that she will get away with
+replacing this code with `Task.Run()`.
+
+## Requirements
+
+### Goals
+
+* The analyzer should be on default and help developers targeting WebAssembly to
+  find places where they rely on thread creation but don't run in a browser that
+  supports threads.
+
+### Non-Goals
+
+* [Platform Compatibility Analyzer][platform-compat]. For some annotations we
+  might choose to move to the capability based model, but we generally want to
+  keep the platform analyzer to represent platform specific APIs. In general,
+  `[UnsupportedOSPlatform]` are the most likely candidates for capability based
+  annotation.
+
+## Design
+
+### Analyzer IDs and Severity
+
+Similar to the [Platform Compatibility Analyzer][platform-compat], this analyzer
+should be enabled by default. The IDs should start with CA and use the next
+available slot.
+
+Diagnostic ID | Severity | Description
+--------------|----------|--------------
+CA_____       | Warning  | The API '{0}' requires the capability '{1}'. Either mark the consumer with `{2}` or check by calling '{3}'.
+
+### Capability Attributes
+
+```C#
+namespace System.Diagnostics.CodeAnalysis
+{
+    public abstract class CapabilityAttribute : Attribute
+    {
+        protected CapabilityAttribute();
+    }
+
+    [AttributeUsage(AttributeTargets.Class, AllowMultiple = true, Inherited = true)]
+    public sealed class CapabilityCheckAttribute : Attribute
+    {
+        public CapabilityCheckAttribute(Type type, string propertyName);
+        public Type Type { get; }
+        public string PropertyName { get; }
+    }
+
+    [AttributeUsage(AttributeTargets.Method |
+                    AttributeTargets.Property |
+                    AttributeTargets.Field,
```

Comment:
Do you have an example of which capability would be controlled by field?

Reply:
We have done this for `[UnsupportedOSPlatformGuard]`, for [example here](https://source.dot.net/#System.Runtime.Caching/System/Runtime/Caching/MemoryCache.cs,44). The intent is to cache the result of the call.